vendor/riscv/riscv-isa-sim/riscv/Proc.cc

```diff
@@ -37,33 +59,136 @@ st_rvfi Processor::step(size_t n, st_rvfi reference) {
   rvfi.rs2_addr = this->get_state()->last_inst_fetched.rs2();
   // TODO add rs2_value
 
-  rvfi.trap = this->taken_trap;
-  rvfi.trap |= (this->which_trap << 1);
 
+  if(this->next_rvfi_intr){
+    rvfi.intr = next_rvfi_intr;
+    this->next_rvfi_intr = 0;
+
+    //Add csr changes that happened during first interrupt step
+    reg_commits.insert(prev_commit_log_reg.begin(), prev_commit_log_reg.end());
+  }
+  
+  // Output dbg caused from EBREAK the previous instruction
+  if(this->next_debug) {
+    rvfi.dbg = this->next_debug;
+    this->next_debug = 0;
+  }
+
+  if(this->state.debug_mode  && (prev_state.debug_mode == 0)){
+    // New external debug request
+    if((this->halt_request != HR_NONE)){ 
+      rvfi.dbg = this->get_state()->dcsr->cause;
+      rvfi.dbg_mode = 1;
+
+    // EBREAK
+    } else if(this->get_state()->dcsr->cause == DCSR_CAUSE_SWBP) {
+      // EBREAK will set debug_mode to 1, but we should report this at the next instruction
+      rvfi.trap |= 1 << 0; //trap [0]
+      rvfi.trap |= 1 << 2; //debug [2]
+      rvfi.trap |= 0xE00 & ((DCSR_CAUSE_SWBP) << 9); //debug cause [11:9]
+      
+      this->next_debug = DCSR_CAUSE_SWBP;
+    }
+  }
+
+  // Set dbg_mode to 1 the first instruction in debug mode, but delay turning 
+  // off dbg_mode to the next instruction after turning off to keep dbg_mode on for dret
+  if( (this->halt_request != HR_NONE)  && (prev_state.debug_mode == 0)) {
+    rvfi.dbg_mode = 1;
+  } else {
+    rvfi.dbg_mode = prev_state.debug_mode;
+  }
+
+
+  if(this->taken_trap) {
+    //interrrupts are marked with the msb high in which_trap
+    if(this->which_trap & ((reg_t)1 << (isa->get_max_xlen() - 1))) { 
+      //Since spike steps two times to take an interrupt, we store the intr value to the next step to return with rvfi
+      this->next_rvfi_intr |= 1 << 0; //intr [0]
+      this->next_rvfi_intr |= 1 << 2; //interrupt [2]
+      this->next_rvfi_intr |= 0x3FF8 & ((this->which_trap & 0xFF) << 3); //cause[13:3]
+    } else{
+      rvfi.trap |= 1 << 0; //trap [0]
+      rvfi.trap |= 1 << 1; //exception [1]
+      rvfi.trap |= 0x1F8 & ((this->which_trap) << 3); //exception_cause [8:3]
+      //TODO:
```

Comment:
Just flagging all new TODOs in review comments, for visibility.

Reply:
The `rvfi_trap` signal is not fully implemented for debug mode like [this](https://docs.openhwgroup.org/projects/cv32e40s-user-manual/en/latest/rvfi.html#rvfi-trap-type).